synapse/storage/data_stores/main/ui_auth.py

```diff
@@ -0,0 +1,294 @@
+        while attempts < 5:
+            session_id = stringutils.random_string(24)
+
+            try:
+                await self.db.simple_insert(
+                    table="ui_auth_sessions",
+                    values={
+                        "session_id": session_id,
+                        "clientdict": clientdict_json,
+                        "uri": uri,
+                        "method": method,
+                        "description": description,
+                        "serverdict": "{}",
+                        "creation_time": self.hs.get_clock().time_msec(),
+                    },
+                    desc="create_ui_auth_session",
+                )
+                return UIAuthSessionData(
+                    session_id, clientdict, uri, method, description
+                )
+            except self.db.engine.module.IntegrityError:
+                attempts += 1
+        raise StoreError(500, "Couldn't generate a session ID.")
+
+    async def get_ui_auth_session(self, session_id: str) -> UIAuthSessionData:
+        """Retrieve a UI auth session.
+
+        Args:
+            session_id: The ID of the session.
+        Returns:
+            A dict containing the device information.
+        Raises:
+            SynapseError: if the session is not found.
+        """
+        try:
+            result = await self.db.simple_select_one(
+                table="ui_auth_sessions",
+                keyvalues={"session_id": session_id},
+                retcols=("clientdict", "uri", "method", "description"),
+                desc="get_ui_auth_session",
+            )
+        except StoreError:
+            raise SynapseError(400, "Unknown session ID: %s" % session_id)
+
+        result["clientdict"] = json.loads(result["clientdict"])
+
+        return UIAuthSessionData(session_id, **result)
+
+    async def mark_ui_auth_stage_complete(
+        self, session_id: str, stage_type: str, result: Union[str, bool, JsonDict],
+    ):
+        """
+        Mark a session stage as completed.
+
+        Args:
+            session_id: The ID of the corresponding session.
+            stage_type: The completed stage type.
+            result: The result of the stage verification.
+        """
+        await self.db.runInteraction(
+            "mark_ui_auth_stage_complete",
+            self._mark_ui_auth_stage_complete,
+            session_id,
+            stage_type,
+            result,
+        )
+
+    def _mark_ui_auth_stage_complete(
+        self, txn, session_id: str, stage_type: str, result: Union[str, bool, JsonDict],
+    ):
+        try:
+            # Get the session to ensure it exists.
+            self.db.simple_select_one_txn(
+                txn,
+                table="ui_auth_sessions",
+                keyvalues={"session_id": session_id},
+                retcols=("session_id",),
+            )
+        except StoreError:
+            raise SynapseError(400, "Unknown session ID: %s" % session_id)
```

Comment:
I'm always a bit dubious about functions in the depths of the storage layer deciding whether it's appropriate to return a 400 or a 401 or any other HTTP error code. we might not even have got here via a REST call. It might be the least-awful option in this case, but could you at least document it in the docstring for `mark_ui_auth_stage_complete` ?

Reply:
Is there a better option to raising a `SynapseError` here? Do we generally create custom exceptions and raise those? I could just let the `StoreError` fall through to the REST layer as well.

Reply:
letting the StoreError propagate seems reasonable, especially if you document it. (Ironically, `StoreError` extends `SynapseError`. I wish it didn't. But at least it's an exception type that can be documented and caught further up the stack.).

Creating a custom exception type doesn't seem a terrible idea either.

Reply:
I just passed the `StoreError` for each of these functions through, it seemed to match how we handle this in other places.